samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c

```diff
@@ -0,0 +1,824 @@
+		}
+		LOG_DBG("MQTT client connected!");
+		break;
+
+	case MQTT_EVT_DISCONNECT:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT disconnect failed %d", evt->result);
+			break;
+		}
+		LOG_DBG("MQTT client disconnected %d\n", evt->result);
+		ctx.connected = false;
+		break;
+
+	case MQTT_EVT_PUBLISH:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT Publish event failed %d", evt->result);
+			break;
+		}
+		handle_mqtt_publish_evt(c, evt);
+		break;
+
+	case MQTT_EVT_PUBACK:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT PUBACK error %d\n", evt->result);
+			break;
+		}
+		LOG_DBG("PUBACK packet id: %u\n",
+				evt->param.puback.message_id);
+		break;
+
+	case MQTT_EVT_PUBREC:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT PUBREC error %d", evt->result);
+			break;
+		}
+		LOG_DBG("PUBREC packet id: %u", evt->param.pubrec.message_id);
+		{
+			struct mqtt_pubrel_param param = {
+				.message_id = evt->param.pubrel.message_id
+			};
+			ret = mqtt_publish_qos2_release(&client, &param);
+			if (ret) {
+				LOG_ERR("mqtt_publish_qos2_release: Failed! %d",
+					ret);
+			} else {
+				LOG_DBG("Release, id %u",
+					evt->param.pubrec.message_id);
+			}
+		}
+		break;
+
+	case MQTT_EVT_PUBREL:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT PUBREL error %d", evt->result);
+			break;
+		}
+		LOG_DBG("PUBREL packet id %u", evt->param.pubrel.message_id);
+		{
+			struct mqtt_pubcomp_param param = {
+				.message_id = evt->param.pubrel.message_id
+			};
+			ret = mqtt_publish_qos2_complete(&client, &param);
+			if (ret) {
+				LOG_ERR("mqtt_publish_qos2_complete Failed:%d",
+					ret);
+			} else {
+				LOG_DBG("Complete, id %u",
+					evt->param.pubrel.message_id);
+			}
+		}
+		break;
+
+	case MQTT_EVT_PUBCOMP:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT PUBCOMP error %d", evt->result);
+			break;
+		}
+		LOG_DBG("PUBCOMP packet id %u", evt->param.pubcomp.message_id);
+		break;
+
+	case MQTT_EVT_SUBACK:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT SUBACK error %d\n", evt->result);
+			break;
+		}
+		LOG_DBG("SUBACK packet id: %u\n",
+				evt->param.suback.message_id);
+		break;
+
+	default:
+		LOG_DBG("default: %d\n", evt->type);
+		break;
+	}
+}
+
+static void mqtt_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int err;
+
+	while (1) {
+		/* Don't go any further until MQTT is connected */
+		k_sem_take(&mqtt_connected, K_FOREVER);
+		while (ctx.connected) {
+			if (fds.fd == INVALID_FDS) {
+				LOG_WRN("ERROR: no socket to poll");
+				continue;
+			}
+			err = poll(&fds, 1, mqtt_keepalive_time_left(&client));
+			if (err < 0) {
+				LOG_ERR("ERROR: poll %d", errno);
+				continue;
+			}
+			err = mqtt_live(&client);
+			if ((err != 0) && (err != -EAGAIN)) {
+				LOG_ERR("ERROR: mqtt_live %d", err);
+				continue;
+			}
+			if ((fds.revents & POLLIN) == POLLIN) {
+				err = mqtt_input(&client);
+				if (err != 0) {
+					LOG_ERR("ERROR: mqtt_input %d", err);
+					continue;
+				}
+			}
+			if ((fds.revents & POLLERR) == POLLERR) {
+				LOG_ERR("POLLERR");
```

Comment:
Errors detected at poll level (POLLERR, POLNVAL) should shutdown the mqtt context manually (`mqtt_disconnect` or `mqtt_abort`, the latter is likely a better fit here) - in case of such errors, `mqtt_input` will not be called, so the mqtt engine will not be able to shut down on its own. Otherwise, we'll miss the `MQTT_EVT_DISCONNECT` event.

Reply:
mqtt_abort is added to the poll error casees. MQTT_EVT_DISCONNECT is notified after calling mqtt_abort.